### ISSUE TITLE
StreamMap notify on stream closed

### ISSUE THREAD

**Issue by aviramha** [tokio-rs/tokio#4847]
**Is your feature request related to a problem? Please describe.**
I tried using `StreamMap` for our use case, and found that the only thing it lacks is notifying when a `Stream` inside it is closed. We could wrap the original `Stream` using a struct that `impl Drop` but our "drop" function requires async context and would also require to somehow communicate to the main loop that it was dropped. I figured that changing `StreamMap` made more sense but I'm open to other implementations ideas

**Describe the solution you'd like**
I propose to change `StreamMap.next` to return `Option<K,Option<V>>`, and when `Option<K, None>` is returned it means that `K` was closed. I am open to other ideas, such as new method (but would require another trait or impl?)
It can also be a different structs

**Describe alternatives you've considered**
Implementing `Drop` for my stream.

**Additional context**
I have created a crate with the modified struct as a reference [here](https://github.com/metalbear-co/streammap-ext) also to be used by us until this is upstream (hopefully it will be :))

**Comment by carllerche** [comment]
Another option would be to map your stream before inserting it into `StreamMap`. When a stream is closed, you issue a value that signals it is closed.

**Comment by aviramha** [comment]
> Another option would be to map your stream before inserting it into `StreamMap`. When a stream is closed, you issue a value that signals it is closed.

Yes but then any Stream would needed to be wrapped, which is ofc possible but I feel less attractive. I think I'll do a PR to add a method that does it and that's a fairly non-breaking way to achieve something that might be useful for others?

**Comment by sharifhsn** [comment]
Closed by #4851

Repro:

```rust
use tokio_stream::{StreamExt, StreamMap, StreamNotifyClose};

#[tokio::main(flavor = "current_thread")]
async fn main() {
    let mut map = StreamMap::new();
    // Wrap the stream so its close is reported.
    map.insert("a", StreamNotifyClose::new(tokio_stream::iter(vec![1, 2])));

    // Items arrive as (key, Some(v)); the close arrives once as (key, None).
    while let Some((key, val)) = map.next().await {
        println!("{key}: {val:?}");
    }
}
```

tokio-stream 0.1.14:

```
a: Some(1)
a: Some(2)
a: None
```

**Comment by martin-g** [comment]
Thank you, @sharifhsn !
